### PR TITLE
Modelling Group and User Auth Service

### PR DESCRIPTION
**Pull request by daudich**

### CODE REVIEW

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+        context.`,
+      type: 'foam.nanos.auth.User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'user',
+          of: 'foam.nanos.auth.User'
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:210]: Should be type:

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+        and group enabled flags.`,
+      type: 'boolean',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'user',
+          of: 'foam.nanos.auth.User'
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:327]: Should be type:

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        },
+        {
+          name: 'permission',
+          of: 'java.security.Permission'
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:331]: should be javaType:

**Comment by adamvy** [comment]
This doesn't compile.

Needs to be added to classes.js and there's a test in NANOPAY that needs to be updated.

I imagine you're not doing a clean build so your code is still picking up compiled class from the hand written code.

**Comment by adamvy** [comment]
Once you do add it to classes.js you'll notice a bunch of new warnings when generating java code from the usage of "of:" instead of type/javaType.  Be sure to fix all those.

**Pull request merged**